73_ArithmeticBench/main.cpp

```diff
@@ -0,0 +1,949 @@
+		const auto dataFromBuffer = reinterpret_cast<const uint32_t*>(resultsBuffer->getPointer());
+		const auto subgroupSize = dataFromBuffer[0];
+		if (subgroupSize<nbl::hlsl::subgroup::MinSubgroupSize || subgroupSize>nbl::hlsl::subgroup::MaxSubgroupSize)
+		{
+			m_logger->log("Unexpected Subgroup Size %u", ILogger::ELL_ERROR, subgroupSize);
+			return false;
+		}
+
+		const auto testData = reinterpret_cast<const type_t*>(dataFromBuffer + 1);
+		// TODO: parallel for (the temporary values need to be threadlocal or what?)
+		// now check if the data obtained has valid values
+		type_t* tmp = new type_t[itemsPerWG * itemsPerInvoc];
+		//type_t* ballotInput = new type_t[itemsPerWG];
+		for (uint32_t workgroupID = 0u; success && workgroupID < workgroupCount; workgroupID++)
+		{
+			const auto workgroupOffset = workgroupID * itemsPerWG * itemsPerInvoc;
+
+			//if constexpr (WorkgroupTest)
+			//{
+			//	if constexpr (std::is_same_v<ballot<type_t>, Binop>)
+			//	{
+			//		for (auto i = 0u; i < itemsPerWG; i++)
+			//			ballotInput[i] = inputData[i + workgroupOffset] & 0x1u;
+			//		Arithmetic<Binop>::impl(tmp, ballotInput, itemsPerWG);
+			//	}
+			//	else
+			//		Arithmetic<Binop>::impl(tmp, inputData + workgroupOffset, itemsPerWG);
+			//}
+			//else
+			//{
+				for (uint32_t pseudoSubgroupID = 0u; pseudoSubgroupID < itemsPerWG; pseudoSubgroupID += subgroupSize)
+					Arithmetic<Binop>::impl(tmp + pseudoSubgroupID * itemsPerInvoc, inputData + workgroupOffset + pseudoSubgroupID * itemsPerInvoc, subgroupSize * itemsPerInvoc);
+			//}
+
+			for (uint32_t localInvocationIndex = 0u; localInvocationIndex < itemsPerWG; localInvocationIndex++)
+			{
+				const auto localOffset = localInvocationIndex * itemsPerInvoc;
+				const auto globalInvocationIndex = workgroupOffset + localOffset;
+
+				for (uint32_t itemInvocationIndex = 0u; itemInvocationIndex < itemsPerInvoc; itemInvocationIndex++)
+				{
+					const auto cpuVal = tmp[localOffset + itemInvocationIndex];
+					const auto gpuVal = testData[globalInvocationIndex + itemInvocationIndex];
+					if (cpuVal != gpuVal)
+					{
+						m_logger->log(
+							"Failed test #%d  (%s)  (%s) Expected %u got %u for workgroup %d and localinvoc %d and iteminvoc %d",
+							ILogger::ELL_ERROR, itemsPerWG, WorkgroupTest ? "workgroup" : "subgroup", Binop::name,
+							cpuVal, gpuVal, workgroupID, localInvocationIndex, itemInvocationIndex
+						);
+						success = false;
+						break;
+					}
+				}
+			}
+		}
+		//delete[] ballotInput;
+		delete[] tmp;
+
+		return success;
+	}
+
+
+	template<template<class> class Arithmetic>
+	bool runBenchmark(IGPUCommandBuffer* cmdbuf, const BenchmarkSet& set, const uint32_t elementCount, const uint8_t subgroupSizeLog2)
+	{
+		const uint32_t workgroupCount = elementCount / (set.workgroupSize * set.itemsPerInvocation);
+
+		cmdbuf->bindComputePipeline(set.pipeline.get());
+		cmdbuf->bindDescriptorSets(EPBP_COMPUTE, set.pipeline->getLayout(), 0u, 1u, &benchDs.get());
```

Comment:
the set you can bind once at start of commandbuffer if you never change the layout